text_classification.py

```diff
@@ -0,0 +1,487 @@
+import os
+import re
+import torch
+from torchtext.utils import download_from_url, extract_archive
+from torchtext.data.utils import generate_ngrams
+from torchtext.vocab import build_vocab
+import random
+from torchtext import data
+from tqdm import tqdm
+from torchtext.data import dataset
+from torchtext.data.iterator import generate_iterators
+
+
+class TextDataset(torch.utils.data.Dataset):
+    def __init__(self, examples, fields):
+        self.examples = examples
+        self.fields = fields
+
+    def __getitem__(self, i):
+        return self.examples[i]
+
+    def __len__(self):
+        try:
+            return len(self.examples)
+        except TypeError:
+            return 2**32
+
+    def __iter__(self):
+        for x in self.examples:
+            yield x
+
+
+class TextClassificationDataset(torch.utils.data.Dataset):
+    """Defines text classification datasets.
+        Currently, we only support the following datasets:
+
+             - AG_NEWS
+             - SogouNews
+             - DBpedia
+             - YelpReviewPolarity
+             - YelpReviewFull
+             - YahooAnswers
+             - AmazonReviewPolarity
+             - AmazonReviewFull
+
+    """
+
+    def __init__(self, dataset_name, root='.data',
+                 text_field=None, label_field=None, ngrams=1):
```

Comment:
The number of ngrams is passed in as a parameter for load_text_classification_data, which in turn calls into label_text_processor and returns the preprocessed "examples" (corpus). build_vocab then takes an instance of self.  I agree that using the iterator approach is more efficient, but in this instance it is iterating over a corpus that's already entirely in memory. So, when build the vocabulary it should be enough to only pass in self.examples. Then build_vocab could stand entirely by itself (maybe even called "build_vocab_from_list") .